### ISSUE TITLE
[Bug]: Issue with setting axis limits on 3D plots

### ISSUE THREAD

**Issue by nbfazel** [matplotlib/matplotlib#25804]
### Bug summary

I have created 3D surface plots using matplotlib as described [[here](https://matplotlib.org/stable/gallery/mplot3d/subplot3d.html)]. Everything works except when I set the axis limits. Then the plots look very strange. It appears that commands such as Axes.set_xlim() (whether or not Axes.set_xbound()is used) limit the axis but not the data. The data is plotted and appears to continue beyond the axis limits.

### Code for reproduction

```python
axes = []
fig0, ax0 = plt.subplots(subplot_kw={"projection": "3d"})
axes.append(ax0)

surf0 = axes[0].plot_surface(DS_ee, DS_cc, DS_array, cmap=cm.jet, linewidth=0, antialiased=False)

axes[0].xaxis.set_major_locator(MultipleLocator(10))
axes[0].xaxis.set_minor_locator(MultipleLocator(2))
axes[0].yaxis.set_major_locator(MultipleLocator(1))
axes[0].yaxis.set_minor_locator(MultipleLocator(0.2))

axes[0].set_zlabel('Default Spectrum [photons/keV/pixel]', fontsize=7)
axes[0].set_xlim(left=0, right=10)
axes[0].set_zlim(0, 3000)
# axes[0].set_xbound(0, 10)
# axes[0].set_zbound(0, 3000)

for t in axes[0].xaxis.get_major_ticks(): t.label.set_fontsize(6)
for t in axes[0].yaxis.get_major_ticks(): t.label.set_fontsize(6)
for t in axes[0].zaxis.get_major_ticks(): t.label.set_fontsize(6)

axes[0].grid(False)
axes[0].view_init(elev=elev_angle, azim=azim_angle)
axes[0].set(xlabel='E [keV]', ylabel=ylabel)
```

### Actual outcome

Surface plot after using `Axes.set_xlim(left=0, right=10)`:

<img width="528" alt="Screenshot 2023-04-28 at 2 53 45 PM" src="https://user-images.githubusercontent.com/14204540/235541107-0d3e0e92-dd65-424f-a985-5dd15e71bc80.png">

Surface plot after using using `Axes.set_xlim(left=0, right=10)` and `axes[0].set_zlim(0, 3000)`: 

<img width="528" alt="Screenshot 2023-04-28 at 2 54 18 PM" src="https://user-images.githubusercontent.com/14204540/235541190-040cae0f-a3de-4ded-8646-d422963e47da.png">

### Expected outcome

The image should look like the following but with the E axis ending at 10 keV and the vertical axis ending at 3000:

<img width="528" alt="Screenshot 2023-04-28 at 2 51 16 PM" src="https://user-images.githubusercontent.com/14204540/235541426-c979f5fe-95dd-4ea5-91b3-1dc963a37af6.png">

### Additional information

I'm running the code that imports matplotlib in Spyder. I can share upon request the full code and input files needed to reproduce the error.

I tried all the suggestions in [this post](https://stackoverflow.com/questions/17734587/why-is-set-xlim-not-setting-the-x-limits-in-my-figure) and also [this one](https://stackoverflow.com/questions/48986956/correctly-setting-the-axes-limits-in-3d-plots) and none worked. For example, I tried:

```
axes[0].set_xlim(left=0, right=10)
axes[0].set_zlim(0, 3000)
```

Also

```
axes[0].set_xlim(min=0, max=10)
axes[0].set_zlim(0, 3000)
```
and

```
axes[0].set_xlim(left=0, right=10)
axes[0].set_zlim(0, 3000)
axes[0].set_xbound(0, 10)
axes[0].set_zbound(0, 3000)
```
and

```
axes[0].set_xlim3d(left=0, right=10)
axes[0].set_zlim3d(0, 3000)
```

### Operating system

OS/X Ventura 13.3.1

### Matplotlib Version

3.5.2

### Matplotlib Backend

Qt5Agg

### Python version

3.9.13

### Installation

None

**Comment by scottshambaugh** [comment]
I think that the axes not "truncating" the plotted surfaces is the intended behavior, though an option to crop data to the box inside the axis planes would be useful. Right now `clip_box` only clips the 2D projection, and I think `clip_box_3d` or something similar would be a worthwhile feature. 

Somewhat related to https://github.com/matplotlib/matplotlib/issues/23450, where the issue is that we want to be clipping based on the 3d box, not the 2d one.

**Comment by scottshambaugh** [comment]
A workaround in your case might be to edit your data so that points higher than 10 / 3000 on the respective axes are set to `np.nan`. This may require calculating the interpolated contour line at 3000 and manually adding those points to your dataset, so you don't get a jagged "sawtooth" there.

**Comment by nbfazel** [comment]
Thanks. I really don't understand why this is the intended behavior because it sure doesn't make sense and I wasted a lot of time trying to change the behavior. I just want to be able to set the axis limit without modifying the data. Any chance this could be fixed?

**Comment by scottshambaugh** [comment]
Hi @nbfazel thank you for raising the issue! I'll try to take a stab at this, but fair warning that it might not be fast and matplotlib only releases a new version every 6-12 months or so.

I see two ways to implement this:

1) Extend bounding boxes to 3D, which I think would be a good bit more work but is a little more flexible.
2) Have a `clip_to_axis_bounds` flag which does specifically this behavior.

I think (2) is definitely much easier and covers most of the desired use cases here, and could always be a special case of (1) if that ever gets implemented.

**Comment by nbfazel** [comment]
Thank you @scottshambaugh. I appreciate that. I will look for a interim solution while this is being implemented.

**Comment by scottshambaugh** [comment]
I did some digging into this and think it will end up being rather complicated for surfaces plots like your example. The problem is that matplotlib's 3D renderer does not use a ray-tracing algorithm, so you cannot easily threshold drawing the face for a polygon between specific bounds.

Take (A) as your initial set of points. It would be fairly easy to threshold your points / vertices based on bounds, and I think that would work well for scatter plots. In the case of surfaces however, this will likely mean jagged sawtooth edges since you would not be drawing the polygons which have a vertex right outside your bounds, as in (B). Perhaps that's okay as a first step implementation. 

The way to get around this would be to fill in the gaps with new polygons created on the fly to edge match with the bounds, as in (C). That however could get very complicated especially in 3D near the corners/edges of the bounds from multiple axes, or in cases where your bounds are tighter than the extent of a single polygon.

![image](https://github.com/matplotlib/matplotlib/assets/14363975/c8f3a180-1a14-4d00-a5af-1dadf3b12c91)

**Comment by nbfazel** [comment]
Thanks for the update. I wonder how this is done Matlab or Mathematica.

**Comment by nbfazel** [comment]
In [this example](https://stackoverflow.com/questions/42924258/create-sea-surface-plot-with-matplotlib3d-wireframe) (scroll down to the answer), limits are set for all axes, without running into the issues I described above. I thought maybe it was because it also set axes to off. I tried that and it didn't make a difference. I'm not sure why setting axis limits works in the example. (I ran the example and produce the same image.)

**Comment by nbfazel** [comment]
Just checking to see if we know why in the example provided above (on May 21), they don't run into the problem after setting axis limit in 3D. Thanks!

**Comment by scottshambaugh** [comment]
The issue persists in the example above. If you don't turn the axes off, then this is more visible:
![image](https://github.com/matplotlib/matplotlib/assets/14363975/3d0a1c1f-20f7-4a72-b10d-710146e49fc6)

If you zoom out, you can see that the limits weren't truncated for the data:
![image](https://github.com/matplotlib/matplotlib/assets/14363975/9be2fb05-ce6d-44ed-bc4f-4dc1ef33635d)

**Comment by nbfazel** [comment]
Thank you so much. So do you think I should just go with the advice you gave on May 2? ("edit your data so that points higher than 10 / 3000 on the respective axes are set to np.nan")

**Comment by scottshambaugh** [comment]
If you need to make these plots any time soon, then yes. With matplotlib as an all-volunteer effort it can take quite a long time to implement feature requests, and the number of people familiar with the 3D code is a good bit less than the 2D plotting. If you want to take a shot at this though, I think that clipping the data to the axis limits would be a welcome feature. :)

**Comment by nbfazel** [comment]
Thanks! Recently I tried to [fix an issue with SciPy](https://github.com/scipy/scipy/issues/17133#issuecomment-1411406925) (that I had reported) and even though the fix wasn't complicated, I spent a lot of time trying to set up my development environment and ultimately gave up. This is more challenging since I would first need to familiarize myself with 3D rendering and then again try to set up my development environment.

**Comment by harish-srini** [comment]
LIke the suggestion by @scottshambaugh what if matplotlib chose the make the plots using the _copy of the original numpy array_ with data points set to np.nan based on the choice of limits? Would that be easy to implement within matplotlib itself?

**Comment by scottshambaugh** [comment]
Draft PR is up, I'll be fleshing it out over the next week. So that PR will implement option (B) from [my comment above](https://github.com/matplotlib/matplotlib/issues/25804#issuecomment-1540429601), with jagged edges at the boundary. I'm going to say that getting straight/smooth edges rather than jagged edges is a won't do, as it would require some pretty drastic changes to how the renderer works, so (C) won't happen and I'll close this issue once the PR is merged.

**Comment by nbfazel** [comment]
OK, thank you. I'll give it a try when it becomes available.

**Comment by scottshambaugh** [comment]
https://github.com/matplotlib/matplotlib/pull/27349 to implement dynamic clipping in 3D plots is now ready for review if you'd like to play with it!

**Comment by nbfazel** [comment]
Thanks! What do I need to do to sync up and pull the version of the code with the fix?

**Comment by scottshambaugh** [comment]
@nbfazel you can pull down the beach that is getting merged here: https://github.com/scottshambaugh/matplotlib/tree/3d_dynamic_masking

I’m on mobile so don’t have the right command to add a branch from a fork of a repo to your local machine, but it should be a pretty easy google.
